src/tracing/metric_macros.h

```diff
@@ -0,0 +1,283 @@
+ */
+#define MATTER_LOG_METRIC(key, ...) __MATTER_LOG_METRIC(chip::Tracing::MetricEvent::Type::kInstantEvent, key, ##__VA_ARGS__)
+
+/**
+ * @def MATTER_LOG_METRIC_BEGIN
+ *
+ * @brief
+ * Generate a metric with the Begin Type
+ *
+ *  Example usage:
+ *  @code
+ *      MATTER_LOG_METRIC_BEGIN(chip::Tracing::kMetricPASESession);
+ *  @endcode
+ *      The above example generates a Begin metric event with key kMetricPASESession.
+ *
+ *  @param[in]  key The key representing the metric name/event.
+ *
+ *  @param[in]  value An optional value for the metric. This value corresponds to one of the values supported
+ *                    in MetricEvent::Value
+ */
+#define MATTER_LOG_METRIC_BEGIN(key, ...) __MATTER_LOG_METRIC(chip::Tracing::MetricEvent::Type::kBeginEvent, key, ##__VA_ARGS__)
+
+/**
+ * @def MATTER_LOG_METRIC_END
+ *
+ * @brief
+ * Generate a metric with the End Type
+ *
+ *  Example usage:
+ *  @code
+ *      MATTER_LOG_METRIC_END(chip::Tracing::kMetricPASESession);
+ *  @endcode
+ *      The above example generates an End metric event with key kMetricPASESession.
+ *
+ *  @param[in]  key The key representing the metric name/event.
+ *
+ *  @param[in]  value An optional value for the metric. This value corresponds to one of the values supported
+ *                    in MetricEvent::Value
+ */
+#define MATTER_LOG_METRIC_END(key, ...) __MATTER_LOG_METRIC(chip::Tracing::MetricEvent::Type::kEndEvent, key, ##__VA_ARGS__)
+
+/**
+ * @def MATTER_LOG_METRIC_SCOPE
+ *
+ * @brief
+ * Generate a scoped metric tracking Begin and End within a given scope.
+ *
+ *  Example usage:
+ *  @code
+ *      MATTER_LOG_METRIC_SCOPE(chip::Tracing::kMetricPASESession);
+ *  @endcode
+ *      The above example generates an Begin and the End metric using RAII.
+ *
+ *  @param[in]  key The key representing the metric name/event.
+ */
+#define MATTER_LOG_METRIC_SCOPE(key)                                                                                               \
+    ::chip::Tracing::utils::ScopedMetricEvent __LOG_METRIC_MACRO_CONCAT(_metric_scope, __COUNTER__)(key)
+
+/**
+ * @def MATTER_LOG_METRIC_SCOPE_WITH_ERROR
+ *
+ * @brief
+ * Generate a scoped metric tracking Begin and End within a given scope. In addition, it creates the object using
+ * the name specified. This object is meant to be used as ChipError object. The End metric will also hold this error
+ * value.
+ *
+ *  Example usage:
+ *  @code
+ *      MATTER_LOG_METRIC_SCOPE_WITH_ERROR(chip::Tracing::kMetricPASESession, err, CHIP_NO_ERROR);
+ *  @endcode
+ *      The above example generates an Begin and the End metric using RAII.
+ *
+ *  @param[in]  key The key representing the metric name/event.
+ *  @param[in]  errorObj The name of the object for the ScopedMetricEvent.
+ *  @param[in]  errorValue The initial error code value.
+ */
+#define MATTER_LOG_METRIC_SCOPE_WITH_ERROR(key, errorObj, errorValue)                                                              \
```

Comment:
I would drop this macro completely then as it provides just hiding of things instead of convenience. The other TRACE provides an `automatic unique name` hence the macro.

However in this case we should just replace `CHIP_ERROR err` with `TracedError err(key)`. This removes error_value (it probably should always be CHIP_NO_ERROR unless we construct with something else).

Reply:
What would the code look like when tracing is disabled? We need the code to be least intrusive. So we need some macro